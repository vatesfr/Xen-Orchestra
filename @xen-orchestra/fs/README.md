<!-- DO NOT EDIT MANUALLY, THIS FILE HAS BEEN GENERATED -->

# @xen-orchestra/fs [![Build Status](https://travis-ci.org/vatesfr/xen-orchestra.png?branch=master)](https://travis-ci.org/vatesfr/xen-orchestra)

[![Package Version](https://badgen.net/npm/v/@xen-orchestra/fs)](https://npmjs.org/package/@xen-orchestra/fs) ![License](https://badgen.net/npm/license/@xen-orchestra/fs) [![PackagePhobia](https://badgen.net/packagephobia/install/@xen-orchestra/fs)](https://packagephobia.now.sh/result?p=@xen-orchestra/fs)

> The File System for Xen Orchestra backups.

## Install

Installation of the [npm package](https://npmjs.org/package/@xen-orchestra/fs):

```
> npm install --global @xen-orchestra/fs
```

## Contributions

Contributions are _very_ welcomed, either on the documentation or on
the code.

You may:

- report any [issue](https://github.com/vatesfr/xen-orchestra/issues)
  you've encountered;
- fork and create a pull request.

## License

[AGPL-3.0-or-later](hhttps://spdx.org/licenses/AGPL-3.0-or-later) Â© [Vates SAS](https://vates.fr)
