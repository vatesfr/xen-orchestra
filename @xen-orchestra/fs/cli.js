#!/usr/bin/env node

'use strict'

const Disposable = require('promise-toolbox/Disposable')
const { getBoundPropertyDescriptor } = require('bind-property-descriptor')

const { getSyncedHandler } = require('./')

const { getPrototypeOf, ownKeys } = Reflect
function getAllBoundDescriptors(object) {
  const descriptors = { __proto__: null }
  let current = object
  do {
    ownKeys(current).forEach(key => {
      if (!(key in descriptors)) {
        descriptors[key] = getBoundPropertyDescriptor(current, key, object)
      }
    })
  } while ((current = getPrototypeOf(current)) !== null)
  return descriptors
}

// https://gist.github.com/julien-f/18161f6032e808d6fa08782951ce3bfb
async function repl({ prompt, context } = {}) {
  const repl = require('repl').start({
    ignoreUndefined: true,
    prompt,
  })
  if (context !== undefined) {
    Object.defineProperties(repl.context, Object.getOwnPropertyDescriptors(context))
  }
  const { eval: evaluate } = repl
  repl.eval = (cmd, context, filename, cb) => {
    evaluate.call(repl, cmd, context, filename, (error, result) => {
      if (error != null) {
        return cb(error)
      }
      Promise.resolve(result).then(result => cb(undefined, result), cb)
    })
  }
  return new Promise((resolve, reject) => {
    repl.on('error', reject).on('exit', resolve)
  })
}

async function* main([url]) {
  if (url === undefined) {
    throw new TypeError('missing arg <url>')
  }

  const handler = yield getSyncedHandler({ url })
  await repl({
    prompt: handler.type + '> ',
    context: Object.create(null, getAllBoundDescriptors(handler)),
  })
}

Disposable.wrap(main)(process.argv.slice(2)).catch(error => {
  console.error('FATAL:', error)
  process.exitCode = 1
})
