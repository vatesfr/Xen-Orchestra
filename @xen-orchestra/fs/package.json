{
  "name": "@xen-orchestra/fs",
  "version": "0.4.1",
  "license": "AGPL-3.0",
  "description": "The File System for Xen Orchestra backups.",
  "keywords": [],
  "homepage": "https://github.com/vatesfr/xen-orchestra/tree/master/@xen-orchestra/fs",
  "bugs": "https://github.com/vatesfr/xen-orchestra/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/vatesfr/xen-orchestra.git"
  },
  "preferGlobal": true,
  "main": "dist/",
  "bin": {},
  "files": [
    "dist/"
  ],
  "engines": {
    "node": ">=6"
  },
  "dependencies": {
    "@marsaud/smb2": "^0.12.0",
    "@xen-orchestra/async-map": "^0.0.0",
    "execa": "^1.0.0",
    "fs-extra": "^7.0.0",
    "get-stream": "^4.0.0",
    "golike-defer": "^0.4.1",
    "lodash": "^4.17.4",
    "promise-toolbox": "^0.11.0",
    "readable-stream": "^3.0.6",
    "through2": "^3.0.0",
    "tmp": "^0.0.33",
    "xo-remote-parser": "^0.5.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.0.0",
    "@babel/core": "^7.0.0",
    "@babel/plugin-proposal-decorators": "^7.1.6",
    "@babel/plugin-proposal-function-bind": "^7.0.0",
    "@babel/preset-env": "^7.0.0",
    "@babel/preset-flow": "^7.0.0",
    "async-iterator-to-stream": "^1.1.0",
    "babel-plugin-lodash": "^3.3.2",
    "cross-env": "^5.1.3",
    "dotenv": "^6.1.0",
    "index-modules": "^0.3.0",
    "rimraf": "^2.6.2"
  },
  "scripts": {
    "build": "cross-env NODE_ENV=production babel --source-maps --out-dir=dist/ src/",
    "clean": "rimraf dist/",
    "dev": "cross-env NODE_ENV=development babel --watch --source-maps --out-dir=dist/ src/",
    "prebuild": "yarn run clean",
    "predev": "yarn run clean",
    "prepare": "yarn run build"
  }
}
