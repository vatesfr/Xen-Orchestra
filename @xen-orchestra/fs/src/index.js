// @flow
import execa from 'execa'

import type RemoteHandler from './abstract'
import RemoteHandlerLocal from './local'
import RemoteHandlerNfs from './nfs'
import RemoteHandlerSmb from './smb'
import RemoteHandlerSmbMount from './smb-mount'

export type { default as RemoteHandler } from './abstract'
export type Remote = { url: string }

const HANDLERS = {
  file: RemoteHandlerLocal,
  nfs: RemoteHandlerNfs,
}

try {
  execa.sync('mount.cifs', ['-V'])
  HANDLERS.smb = RemoteHandlerSmbMount
} catch (_) {
  HANDLERS.smb = RemoteHandlerSmb
}

export const getHandler = (remote: Remote, ...rest: any): RemoteHandler => {
  // FIXME: should be done in xo-remote-parser.
  const type = remote.url.split('://')[0]

  const Handler = HANDLERS[type]
  if (!Handler) {
    throw new Error('Unhandled remote type')
  }
  return new Handler(remote, ...rest)
}
