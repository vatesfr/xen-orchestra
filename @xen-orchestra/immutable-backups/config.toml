
# how often does the lift immutability script will run to check if 
# some files need to be made mutable 
liftEvery =  1h

# you can add as many remote as you want, if you change the id ( here : remote1)
#[remotes.remote1]
#root = "/mnt/ssd/vhdblock/" # the absolute path of the root of the backup repository
#immutabilityDuration = 7d # mandatory
# optional, default value is false will scan and update the index on start, can be expensive
#rebuildIndexOnStart = true 

# the index path is optional, default in XDG_DATA_HOME, or if this is not set, in ~/.local/share
#indexPath = "/var/lib/" # will add automatically the application name immutable-backup 
