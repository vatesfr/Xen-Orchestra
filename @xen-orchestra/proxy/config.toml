# Vendor config: DO NOT TOUCH!
#
# See sample.config.toml to override.

# This secret string is used to authenticate clients to the API.
#
# It must be defined to a non-empty string for the proxy to run.
authenticationToken = ''

datadir = '/var/lib/xo-proxy/data'
resourceDebounce = '5m'

[api]
keepAliveInterval = 10e3

[backups]
snapshotNameLabelTpl = '[XO Backup {job.name}] {vm.name_label}'

[backups.defaultSettings]
bypassVdiChainsCheck = false
checkpointSnapshot = false
concurrency = 2
deleteFirst = false
exportRetention = 0
fullInterval = 0
offlineBackup = false
offlineSnapshot = false
reportWhen = 'failure'
snapshotRetention = 0
timeout = 0
vmTimeout = 0

# This is a work-around.
#
# See https://github.com/vatesfr/xen-orchestra/pull/4674
maxMergedDeltasPerRun = 2

# Each `http.listen.<name>` entry defines a specific listening configuration for
# the HTTP server.
#
# `<name>` can be freely choosen.
[http.listen.https]

# Generate self-signed certificate if missing
autoCert = true

cert = '/var/lib/xo-proxy/certificate.pem'
key = '/var/lib/xo-proxy/key.pem'
port = 443

[remoteOptions]
mountsDir = '/run/xo-proxy/mounts'

# timeout in milliseconds (set to 0 to disable)
timeout = 600e3

# see https:#github.com/vatesfr/xen-orchestra/issues/3419
# useSudo = false

[xapiOptions]
maxUncoalescedVdis = 1
watchEvents = ['network', 'PIF', 'pool', 'SR', 'task', 'VBD', 'VDI', 'VIF', 'VM']
