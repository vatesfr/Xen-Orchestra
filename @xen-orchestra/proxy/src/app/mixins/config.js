import get from 'lodash/get'
import identity from 'lodash/identity'
import { createLogger } from '@xen-orchestra/log'
import { parseDuration } from '@vates/parse-duration'
import { watch } from 'app-conf'

const { warn } = createLogger('xo:proxy:config')

export default class Config {
  constructor(app, { appDir, appName, config }) {
    this._config = config
    const watchers = (this._watchers = new Set())

    app.hooks.on('start', async () => {
      app.hooks.on(
        'stop',
        await watch({ appDir, appName, ignoreUnknownFormats: true }, (error, config) => {
          if (error != null) {
            return warn(error)
          }

          this._config = config
          watchers.forEach(watcher => {
            watcher(config)
          })
        })
      )
    })
  }

  get(path) {
    const value = get(this._config, path)
    if (value === undefined) {
      throw new TypeError('missing config entry: ' + value)
    }
    return value
  }

  getDuration(path) {
    return parseDuration(this.get(path))
  }

  watch(path, cb) {
    // internal arg
    const processor = arguments.length > 2 ? arguments[2] : identity

    let prev
    const watcher = config => {
      try {
        const value = processor(get(config, path))
        if (value !== prev) {
          prev = value
          cb(value)
        }
      } catch (error) {
        warn('watch', { error, path })
      }
    }

    // ensure sync initialization
    watcher(this._config)

    const watchers = this._watchers
    watchers.add(watcher)
    return () => watchers.delete(watcher)
  }

  watchDuration(path, cb) {
    return this.watch(path, cb, parseDuration)
  }
}
