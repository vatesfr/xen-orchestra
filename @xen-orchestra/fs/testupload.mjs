import fs from 'fs/promises'
import { getSignedUrl } from "@aws-sdk/s3-request-presigner"
import { createHash } from "crypto";
import {
    CompleteMultipartUploadCommand,
    CreateMultipartUploadCommand,
    GetObjectLockConfigurationCommand,
    PutObjectCommand,
    S3Client,
    UploadPartCommand,
  } from '@aws-sdk/client-s3'

  import { NodeHttpHandler } from '@aws-sdk/node-http-handler'
import { Agent as HttpAgent } from 'http'
import { Agent as HttpsAgent } from 'https'
import { parse } from 'xo-remote-parser'
import { join, split } from './dist/path.js'

import guessAwsRegion from './dist/_guessAwsRegion.js'
import { PassThrough } from 'stream'
import { readChunk } from '@vates/read-chunk'
import { pFromCallback } from 'promise-toolbox'

async function v2(url, inputStream){
    const {
        allowUnauthorized,
        host,
        path,
        username,
        password,
        protocol,
        region = guessAwsRegion(host),
      } = parse(url)
    const client = new S3Client({
        apiVersion: '2006-03-01',
        endpoint: `${protocol}://s3.us-east-2.amazonaws.com`,
        forcePathStyle: true,
        credentials: {
          accessKeyId: username,
          secretAccessKey: password,
        },
        region,
        requestHandler: new NodeHttpHandler({
          socketTimeout: 600000,
          httpAgent: new HttpAgent({
            keepAlive: true,
          }),
          httpsAgent: new HttpsAgent({
            rejectUnauthorized: !allowUnauthorized,
            keepAlive: true,
          }),
        }),
      })
      
    const pathParts = split(path)
    const bucket = pathParts.shift()
    const dir = join(...pathParts)

    const command = new CreateMultipartUploadCommand({
        Bucket: bucket, Key: join(dir, 'flov2')
    })
    const multipart = await client.send(command)
    console.log({multipart})

    const parts = []
    // monitor memory usage
    const intervalMonitorMemoryUsage = setInterval(()=>console.log(Math.round(process.memoryUsage().rss/1024/1024)), 2000)

    const CHUNK_SIZE = Math.ceil(5*1024*1024*1024*1024/10000) // smallest chunk allowing 5TB upload

    async function read(inputStream, maxReadSize){
        if(maxReadSize === 0){
            return null
        }
        process.stdout.write('+')
        const chunk = await readChunk(inputStream, maxReadSize)
        process.stdout.write('@')
        return chunk
    }

    async function write(data, chunkStream, remainingBytes){
        const ready = chunkStream.write(data)
        if(!ready){
            process.stdout.write('.')
            await pFromCallback(cb=> chunkStream.once('drain', cb))
            process.stdout.write('@')
        }
        remainingBytes -= data.length
        process.stdout.write(remainingBytes+' ')
        return remainingBytes
    }

    
    async function uploadChunk(inputStream){
        const PartNumber = parts.length +1
        let done = false
        let remainingBytes = CHUNK_SIZE
        const maxChunkPartSize = Math.round(CHUNK_SIZE / 1000)
        const chunkStream = new PassThrough()
        console.log({maxChunkPartSize,CHUNK_SIZE})
        


        let data
        let chunkBuffer = []
        const hash = createHash('md5');
        try{
            while((data = await read(inputStream, Math.min(remainingBytes, maxChunkPartSize))) !== null){
                chunkBuffer.push(data)
                hash.update(data)
                remainingBytes -= data.length
                //remainingBytes  = await write(data, chunkStream, remainingBytes)
            }
            console.log('data put')
            const fullBuffer = Buffer.alloc(maxChunkPartSize,0)
            done = remainingBytes > 0
            // add padding at the end of the file (not a problem for tar like : xva/ova)
            // if not content length will not match and we'll have UND_ERR_REQ_CONTENT_LENGTH_MISMATCH error
            console.log('full padding')
            while(remainingBytes > maxChunkPartSize){
                chunkBuffer.push(fullBuffer)
                hash.update(fullBuffer)
                remainingBytes -= maxChunkPartSize
                //remainingBytes  = await write(fullBuffer,chunkStream, remainingBytes)
            }
            console.log('full padding done ')
            chunkBuffer.push(Buffer.alloc(remainingBytes,0))
            hash.update(Buffer.alloc(remainingBytes,0))
            console.log('md5 ok ')
            //await write(Buffer.alloc(remainingBytes,0),chunkStream, remainingBytes)
            // wait for the end of the upload

            const command = new UploadPartCommand({
                ...multipart,
                PartNumber,
                ContentLength:CHUNK_SIZE,
                Body: chunkStream,
                ContentMD5 : hash.digest('base64')
            })
            const promise = client.send(command)
            for (const buffer of chunkBuffer){
                await write(buffer, chunkStream, remainingBytes)
            }
            chunkStream.on('error', err => console.error(err))
            const res = await  promise

            console.log({res, headers : res.headers })
            parts.push({ ETag:/*res.headers.get('etag')  */res.ETag, PartNumber })
        }catch(err){
            console.error(err)
            throw err
        }
        return done
    }

    while(!await uploadChunk(inputStream)){
        console.log('uploaded one chunk', parts.length)
    }

    // mark the upload as complete and ask s3 to glue the chunk together
    const completRes = await client.send(
        new CompleteMultipartUploadCommand({
          ...multipart,
          MultipartUpload: { Parts: parts },
        })
      )
      console.log({completRes})
    clearInterval(intervalMonitorMemoryUsage)

}

async function simplePut(url , inputStream){
    const {
        allowUnauthorized,
        host,
        path,
        username,
        password,
        protocol,
        region = guessAwsRegion(host),
      } = parse(url)
    const client = new S3Client({
        apiVersion: '2006-03-01',
        endpoint: `${protocol}://s3.us-east-2.amazonaws.com`,
        forcePathStyle: true,
        credentials: {
          accessKeyId: username,
          secretAccessKey: password,
        },
        region,
        requestHandler: new NodeHttpHandler({
          socketTimeout: 600000,
          httpAgent: new HttpAgent({
            keepAlive: true,
          }),
          httpsAgent: new HttpsAgent({
            rejectUnauthorized: !allowUnauthorized,
            keepAlive: true,
          }),
        }),
      })
      
    const pathParts = split(path)
    const bucket = pathParts.shift()
    const dir = join(...pathParts)

    //const hasObjectLock = await client.send(new GetObjectLockConfigurationCommand({Bucket: bucket}))
    //console.log(hasObjectLock.ObjectLockConfiguration?.ObjectLockEnabled === 'Enabled')


    const md5 = await createMD5('/tmp/1g')
    console.log({md5})
    const command = new PutObjectCommand({
        Bucket: bucket, Key: join(dir, 'simple'),
        ContentMD5: md5,
        ContentLength: 1024*1024*1024,
        Body: inputStream
    })
    const intervalMonitorMemoryUsage = setInterval(()=>console.log(Math.round(process.memoryUsage().rss/1024/1024)), 2000)

    const res = await client.send(command)
    /*
    const presignedUrl = await getSignedUrl(client,  command,{ expiresIn: 3600 });
    const res = await fetch(presignedUrl, {
        method: 'PUT',
        body:inputStream,
        duplex: "half",
        headers:{
            "x-amz-decoded-content-length": 1024*1024*1024,
            "content-md5" : md5
        } 
    })*/
    clearInterval(intervalMonitorMemoryUsage)

    console.log(res)
}

async function createMD5(filePath) {
    const input = await fs.open(filePath) // big ass file 
    return new Promise((res, rej) => {
      const hash = createHash('md5');
      
      const rStream = input.createReadStream(filePath);
      rStream.on('data', (data) => {
        hash.update(data);
      });
      rStream.on('end', () => {
        res(hash.digest('base64'));
      });
    })
  }
const input = await fs.open('/tmp/1g') // big ass file 
const inputStream = input.createReadStream()
const remoteUrl = ""

v2(remoteUrl,inputStream)

//simplePut(remoteUrl,inputStream)