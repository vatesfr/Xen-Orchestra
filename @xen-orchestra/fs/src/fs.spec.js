/* eslint-env jest */

import 'dotenv/config'
import asyncIteratorToStream from 'async-iterator-to-stream'
import getStream from 'get-stream'
import { fromCallback } from 'promise-toolbox'
import { pipeline } from 'readable-stream'
import { random } from 'lodash'
import { tmpdir } from 'os'

import { getHandler } from '.'
import fs from 'fs-extra'

// https://gist.github.com/julien-f/3228c3f34fdac01ade09
const unsecureRandomBytes = n => {
  const bytes = Buffer.alloc(n)

  const odd = n & 1
  for (let i = 0, m = n - odd; i < m; i += 2) {
    bytes.writeUInt16BE((Math.random() * 65536) | 0, i)
  }

  if (odd) {
    bytes.writeUInt8((Math.random() * 256) | 0, n - 1)
  }

  return bytes
}

const TEST_DATA_LEN = 1024
const TEST_DATA = unsecureRandomBytes(TEST_DATA_LEN)
const createTestDataStream = asyncIteratorToStream(function*() {
  yield TEST_DATA
})

const rejectionOf = p =>
  p.then(
    value => {
      throw value
    },
    reason => reason
  )

const handlers = [`file://${tmpdir()}`]
if (process.env.xo_fs_nfs) handlers.push(process.env.xo_fs_nfs)
if (process.env.xo_fs_smb) handlers.push(process.env.xo_fs_smb)

handlers.forEach(url => {
  describe(url, () => {
    let handler

    const testWithFileDescriptor = (path, flags, fn) => {
      it('with path', () => fn({ file: path, flags }))
      it('with file descriptor', async () => {
        const file = await handler.openFile(path, flags)
        try {
          await fn({ file })
        } finally {
          await handler.closeFile(file)
        }
      })
    }

    beforeAll(async () => {
      handler = getHandler({ url }).addPrefix(`xo-fs-tests-${Date.now()}`)
      await handler.sync()
    })
    afterAll(async () => {
      await handler.forget()
      handler = undefined
    })

    beforeEach(async () => {
      // ensure test dir exists
      await handler.mkdir('.')
    })
    afterEach(async () => {
      await handler.rmtree('.')
    })

    describe('#type', () => {
      it('returns the type of the remote', () => {
        expect(typeof handler.type).toBe('string')
      })
    })

    describe('#createOutputStream()', () => {
      it('creates parent dir if missing', async () => {
        const stream = await handler.createOutputStream('dir/file')
        await fromCallback(cb => pipeline(createTestDataStream(), stream, cb))
        await expect(await handler.readFile('dir/file')).toEqual(TEST_DATA)
      })
    })

    describe('#createReadStream()', () => {
      beforeEach(() => handler.outputFile('file', TEST_DATA))

      testWithFileDescriptor('file', 'r', async ({ file, flags }) => {
        await expect(
          await getStream.buffer(
            await handler.createReadStream(file, { flags })
          )
        ).toEqual(TEST_DATA)
      })
    })

    describe('#createWriteStream()', () => {
      testWithFileDescriptor('file', 'wx', async ({ file, flags }) => {
        const stream = await handler.createWriteStream(file, { flags })
        await fromCallback(cb => pipeline(createTestDataStream(), stream, cb))
        await expect(await handler.readFile('file')).toEqual(TEST_DATA)
      })

      it('fails if parent dir is missing', async () => {
        const error = await rejectionOf(handler.createWriteStream('dir/file'))
        expect(error.code).toBe('ENOENT')
      })
    })

    describe('#getSize()', () => {
      beforeEach(() => handler.outputFile('file', TEST_DATA))

      testWithFileDescriptor('file', 'r', async () => {
        expect(await handler.getSize('file')).toEqual(TEST_DATA_LEN)
      })
    })

    describe('#list()', () => {
      it(`should list the content of folder`, async () => {
        await handler.outputFile('file', TEST_DATA)
        await expect(await handler.list('.')).toEqual(['file'])
      })

      it('can prepend the directory to entries', async () => {
        await handler.outputFile('dir/file', '')
        expect(await handler.list('dir', { prependDir: true })).toEqual([
          '/dir/file',
        ])
      })

      it('can prepend the directory to entries', async () => {
        await handler.outputFile('dir/file', '')
        expect(await handler.list('dir', { prependDir: true })).toEqual([
          '/dir/file',
        ])
      })
    })

    describe('#mkdir()', () => {
      it('creates a directory', async () => {
        await handler.mkdir('dir')
        await expect(await handler.list('.')).toEqual(['dir'])
      })

      it('does not throw on existing directory', async () => {
        await handler.mkdir('dir')
        await handler.mkdir('dir')
      })

      it('throws ENOTDIR on existing file', async () => {
        await handler.outputFile('file', '')
        const error = await rejectionOf(handler.mkdir('file'))
        expect(error.code).toBe('ENOTDIR')
      })
    })

    describe('#mktree()', () => {
      it('creates a tree of directories', async () => {
        await handler.mktree('dir/dir')
        await expect(await handler.list('.')).toEqual(['dir'])
        await expect(await handler.list('dir')).toEqual(['dir'])
      })

      it('does not throw on existing directory', async () => {
        await handler.mktree('dir/dir')
        await handler.mktree('dir/dir')
      })

      it('throws ENOTDIR on existing file', async () => {
        await handler.outputFile('dir/file', '')
        const error = await rejectionOf(handler.mktree('dir/file'))
        expect(error.code).toBe('ENOTDIR')
      })

      it('throws ENOTDIR on existing file in path', async () => {
        await handler.outputFile('file', '')
        const error = await rejectionOf(handler.mktree('file/dir'))
        expect(error.code).toBe('ENOTDIR')
      })
    })

    describe('#outputFile()', () => {
      it('writes data to a file', async () => {
        await handler.outputFile('file', TEST_DATA)
        expect(await handler.readFile('file')).toEqual(TEST_DATA)
      })

      it('throws on existing files', async () => {
        await handler.outputFile('file', '')
        const error = await rejectionOf(handler.outputFile('file', ''))
        expect(error.code).toBe('EEXIST')
      })
    })

    describe('#read()', () => {
      beforeEach(() => handler.outputFile('file', TEST_DATA))

      const start = random(TEST_DATA_LEN)
      const size = random(TEST_DATA_LEN)

      testWithFileDescriptor('file', 'r', async ({ file }) => {
        const buffer = Buffer.alloc(size)
        const result = await handler.read(file, buffer, start)
        expect(result.buffer).toBe(buffer)
        expect(result).toEqual({
          buffer,
          bytesRead: Math.min(size, TEST_DATA_LEN - start),
        })
      })
    })

    describe('#readFile', () => {
      it('returns a buffer containing the contents of the file', async () => {
        await handler.outputFile('file', TEST_DATA)
        expect(await handler.readFile('file')).toEqual(TEST_DATA)
      })

      it('throws on missing file', async () => {
        const error = await rejectionOf(handler.readFile('file'))
        expect(error.code).toBe('ENOENT')
      })
    })

    describe('#rename()', () => {
      it(`should rename the file`, async () => {
        await handler.outputFile('file', TEST_DATA)
        await handler.rename('file', `file2`)

        expect(await handler.list('.')).toEqual(['file2'])
        expect(await handler.readFile(`file2`)).toEqual(TEST_DATA)
      })
    })

    describe('#rmdir()', () => {
      it('should remove an empty directory', async () => {
        await handler.mkdir('dir')
        await handler.rmdir('dir')
        expect(await handler.list('.')).toEqual([])
      })

      it(`should throw on non-empty directory`, async () => {
        await handler.outputFile('dir/file', '')

        const error = await rejectionOf(handler.rmdir('.'))
        await expect(error.code).toEqual('ENOTEMPTY')
      })

      it('does not throw on missing directory', async () => {
        await handler.rmdir('dir')
      })
    })

    describe('#rmtree', () => {
      it(`should remove a directory resursively`, async () => {
        await handler.outputFile('dir/file', '')
        await handler.rmtree('dir')

        expect(await handler.list('.')).toEqual([])
      })
    })

    describe('#test()', () => {
      it('tests the remote appears to be working', async () => {
        expect(await handler.test()).toEqual({
          success: true,
        })
      })
    })

    describe('#unlink()', () => {
      it(`should remove the file`, async () => {
        await handler.outputFile('file', TEST_DATA)
        await handler.unlink('file')

        await expect(await handler.list('.')).toEqual([])
      })

      it('does not throw on missing file', async () => {
        await handler.unlink('file')
      })
    })
  })

  describe('outputFile()', () => {
    it(`should put the correct content in file`, async () => {
      const content = 'unit test'
      await handler.outputFile(pathFile, content)
      const contentFile = fs.readFileSync(pathFile, 'utf8')

      await expect(contentFile).toEqual(content)
    })
  })

  describe('readFile()', () => {
    it(`should read the correct content in file`, async () => {
      const content = await handler.readFile(pathFile, 'utf8')
      const contentFile = fs.readFileSync(pathFile, 'utf8')

      await expect(content).toEqual(contentFile)
    })
  })

  describe('list()', () => {
    it(`should list the content of folder`, async () => {
      const content = await handler.list(pathFolder)
      const contentFolder = fs.readdirSync(pathFolder)

      await expect(content).toEqual(contentFolder)
    })
  })

  describe('getSize()', () => {
    it(`should return the correct size`, async () => {
      const fileSize = await handler.getSize(pathFile)
      const stats = await fs.statSync(pathFile)

      expect(fileSize).toEqual(stats.size)
    })
  })

  // describe('createReadStream()', () => {
  //   it(`should return correct stream`, async () => {
  //     const stream = await handler.createReadStream(pathFile)
  //     const readStream = fs.createReadStream(pathFile)
  //     await expect(stream).toEqual(readStream)
  //   })
  // })
  // describe('openFile()', () => {
  //   it(`should return the correct file descriptor`, async () => {
  //     const content = await handler.openFile(pathFile, 'r')
  //     const fd = fs.openSync(pathFile, 'r')
  //     await expect(content.fd).toEqual(fd)
  //   })
  // })
})
