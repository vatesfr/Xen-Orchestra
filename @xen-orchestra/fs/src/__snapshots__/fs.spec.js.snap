// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test() handler doesn't crash 1`] = `
Object {
  "success": true,
}
`;
